lightstep/lightstep_b3_propagator.py

```diff
@@ -0,0 +1,127 @@
+                            "be {} instead of {}".format(
+                                int(sampled), sampled
+                            )
+                        )
+        else:
+            if traceid is None or spanid is None:
+                raise Exception(
+                    "If not propagating only the sampling state, traceid and "
+                    "spanid must be defined"
+                )
+
+    def extract(self, carrier):
+        case_insensitive_carrier = {}
+        for key, value in carrier.items():
+            for b3_key in [
+                _SINGLE_HEADER,
+                _TRACEID,
+                _SPANID,
+                _PARENTSPANID,
+                _SAMPLED,
+                _FLAGS,
+            ]:
+                if re_compile(key, IGNORECASE).match(b3_key):
+                    case_insensitive_carrier[b3_key] = value
+        carrier = case_insensitive_carrier
+        if _SINGLE_HEADER in carrier.keys():
+            fields = carrier[_SINGLE_HEADER].split("-", 4)
+            len_fields = len(fields)
+            if len_fields < 3:
+                if len(fields) == 1:
+                    sampled = fields[0]
+                elif len(fields) == 2:
+                    traceid, spanid = fields
+                baggage = None
+            else:
+                baggage = {}
+                if len(fields) == 3:
+                    traceid, spanid, sampled = fields
+                else:
+                    traceid, spanid, sampled, parent_spanid = fields
+                    baggage[_PARENTSPANID] = parent_spanid
+                if sampled == "d":
+                    baggage[_FLAGS] = 1
+                else:
+                    baggage[_SAMPLED] = sampled
+        else:
+            traceid = carrier.get(_TRACEID, None)
+            spanid = carrier.get(_SPANID, None)
+            parentspanid = carrier.get(_PARENTSPANID, None)
+            sampled = carrier.get(_SAMPLED, None)
+            flags = carrier.get(_FLAGS, None)
+            if parentspanid is sampled is flags is None:
+                baggage = None
+            else:
+                baggage = {}
+                if parentspanid is not None:
+                    baggage[_PARENTSPANID] = parentspanid
+                if flags == 1:
+                    baggage[_FLAGS] = flags
+                    if sampled is not None:
+                        warn(
+                            "x-b3-flags: 1 implies x-b3-sampled: 1, ignoring "
+                            "the received value of x-b3-sampled"
+                        )
+                elif sampled is not None:
+                    baggage[_SAMPLED] = sampled
+        return SpanContext(
+            # traceid and spanid are encoded in hex, so thet must be encoded
+            trace_id=int(traceid, 16),
+            span_id=int(spanid, 16),
+            baggage=baggage
+        )
+
+
+__all__ = ['LightStepB3Propagator']
```

Comment:
We don't use this convention, so we don't need it ;)

Reply:
Ok, I'll remove it. How do you specify public interfaces, then?

Reply:
We simply allow users to explicitly load them, i.e.

```python
from lightstep.propagators import B3Propagator
```

(This is a convention we have also kept from OpenTracing)

Reply:
Fixed